### PR TITLE
[Prototype] Implement Storage Entry API

### LINKED ISSUE #379: Implement storage entry API
Currently we have 3 different APIs to access a contract's storage in `ink_core`:

- `env::get_contract_storage`: To retrieve a decoded `T` from the contract storage cell.
- `env::set_contract_storage`: To encode a `T` into the contract storage cell.
- `env::clear_contract_storage`: To clear all data from the contract storage cell.

This API is fine, however, it doesn't scale well for additional operations.
For example, adding the ability to query if a contract storage is vacant or occupied requires adding yet another API to the already large set of root functions.

Instead we could unify all accesses to a single contract storage cell using one entry API:

```
fn storage_entry(at: &Key) -> StorageEntry { ... }


pub struct StorageEntry<'a> {
    key: &'a Key,
}

pub enum StorageEntryState {
    Vacant,
    Occupied,
}

impl StorageEntry<'_> {
    pub fn is_vacant(&self) -> bool {
        ...
    }

    pub fn is_occupied(&self) -> bool {
        ...
    }

    pub fn get<R>(&self) -> Result<Option<R>>
    where
        R: scale::Decode,
    {
        ...
    }

    pub fn set<V>(&self, value: &V)
    where
        V: scale::Encode,
    {
        ...
    }
}
```

This is also easily extensible in case we later add the subtrie API for contracts.

### PR DESCRIPTION
**Pull request by cmichi**
Closes #379.

### CODE REVIEW

File: crates/env/src/engine/on_chain/impls.rs [hunk 1]
```diff
@@ -209,6 +209,15 @@ impl EnvInstance {
 }
 
 impl EnvBackend for EnvInstance {
+    fn is_contract_storage(&mut self, key: &Key) -> bool {
+        let output = &mut self.scoped_buffer().take_rest();
+        match ext::get_storage(key.as_bytes(), output) {
```
Comment (cmichi) [inline crates/env/src/engine/on_chain/impls.rs:214]: A seal operation `is_storage` would do us good here as well, same as in https://github.com/paritytech/ink/issues/507. Maybe it's time?

**Comment by Robbepop** [comment]
Thanks for working on this POC @cmichi !
However, I unfortunately see some problems with this API as such in regards to the notion of "leaky abstractions".
We might not want to merge this feature due to lack of backend support although its interfaces have some charm to themselves.

**Review by Robbepop — COMMENTED** [review]
Will put on ice until we decide how to push forward with this experimental API.

**Comment by cmichi** [comment]
Closing this issue, since we determined to not go down this road. See the comments above for the reasoning.

Nevertheless it was a worthy exploration and we couldn't have come to that conclusion without this PR.

**Pull request closed without merging**